src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -30,153 +29,171 @@ case class PortsMatch(hostPortsWithRole: Seq[Option[PortWithRole]]) {
-      log.info(s"Offer [${offer.getId.getValue}]. $resourceSelector. " +
-        s"Couldn't find $numberOfPorts ports in offer for run spec [${runSpec.id}]")
-      None
+    /**
+      * Choose random ports from offer.
+      */
+    def randomPorts(numberOfPorts: Int): Option[Seq[Option[PortWithRole]]] = {
+      takeEnoughPortsOrNone(expectedSize = numberOfPorts) {
+        shuffledAvailablePorts.map(Some(_))
+      } orElse {
+        log.info(s"Offer [${offer.getId.getValue}]. $resourceSelector. " +
+          s"Couldn't find $numberOfPorts ports in offer for ${name}")
+        None
+      }
     }
-  }
 
-  /**
-    * Try to find all non-zero host ports in offer and use random ports from the offer for dynamic host ports (=0).
-    * Return `None` if not all host ports could be assigned this way.
-    */
-  private[this] def mappedPortRanges(mappings: Seq[PortMapping]): Option[Seq[Option[PortWithRole]]] = {
-    takeEnoughPortsOrNone(expectedSize = mappings.size) {
-      // non-dynamic hostPorts from port mappings
-      val hostPortsFromMappings: Set[Int] = mappings.collect {
-        case PortMapping(_, Some(hostPort), _, _, _, _) if hostPort != 0 => hostPort
-      }.toSet
-
-      // available ports without the ports that have been preset in the port mappings
-      val availablePortsWithoutStaticHostPorts: Iterator[PortWithRole] =
-        shuffledAvailablePorts.filter(portWithRole => !hostPortsFromMappings(portWithRole.port))
-
-      mappings.iterator.map {
-        case PortMapping(containerPort, Some(hostPort), servicePort, protocol, name, labels) if hostPort == 0 =>
-          if (!availablePortsWithoutStaticHostPorts.hasNext) {
-            log.info(s"Offer [${offer.getId.getValue}]. $resourceSelector. " +
-              s"Insufficient ports in offer for run spec [${runSpec.id}]")
-            None
-          } else {
-            Option(availablePortsWithoutStaticHostPorts.next())
-          }
-        case PortMapping(_, Some(hostPort), _, _, _, _) =>
-          offeredPortRanges.find(_.contains(hostPort)) match {
-            case Some(PortRange(role, _, _, reservation)) =>
-              Some(PortWithRole(role, hostPort, reservation))
-            case None =>
+    /**
+      * Try to find all non-zero host ports in offer and use random ports from the offer for dynamic host ports (=0).
+      * Return `None` if not all host ports could be assigned this way.
+      */
+    def mappedPortRanges(mappings: Seq[Mapping]): Option[Seq[Option[PortWithRole]]] = {
+      import PortsMatcher.Mapping
+      takeEnoughPortsOrNone(expectedSize = mappings.size) {
+        // non-dynamic hostPorts from port mappings
+        val hostPortsFromMappings: Set[Int] = mappings.collect {
+          case Mapping(_, Some(hostPort)) if hostPort != 0 => hostPort
+        }.toSet
+
+        // available ports without the ports that have been preset in the port mappings
+        val availablePortsWithoutStaticHostPorts: Iterator[PortWithRole] =
```

Comment:
Seq should be suffcient here. Possibly even a view since you're continuing to manipulate the collection afaict.